core/interop/src/lib.rs

```diff
@@ -305,6 +325,74 @@ impl<'a, T: TryFromJs> TryFromJsArgument<'a> for JsThis<T> {
     }
 }
 
+impl<T: TryFromJs> Deref for JsThis<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Captures a class instance from the `this` value in a JS function. The class
+/// will be a non-mutable reference of Rust type `T`, if it is an instance of `T`.
+///
+/// To have more flexibility on the parsing of the `this` value, you can use the
+/// [`JsThis`] capture instead.
+#[derive(Debug, Clone)]
+pub struct JsClass<T: NativeObject> {
+    inner: boa_engine::JsObject,
```

Comment:
You can use `boa_engine::JsObject<T>` and `JsObject::downcast` instead, which gives us infallible downcasts after the first downcast.

Reply:
Sure, but it downcasts to `Object<T>`, not `T` :(

Reply:
I don't know how to get a `T` (or more specifically a `&mut T`) from an `Object<T>`.

Ideally I want a `GcRef<'_, T>` (and `GcRefMut<'_, _, T>`) directly.

Reply:
Nono, there's a `JsObject::downcast` that converts from `JsObject` to `JsObject<T>, which makes it possible to  access the data between borrows without having to call `downcast_mut` or `downcast_ref`

Reply:
Yes but that data is non-mut. Are you suggesting I should add this to `Object<T>`?

```rust
pub fn data_mut(&mut self) -> &mut T {
  self.data
}
```

Reply:
After that, if you want to map to the inner data, you can do

`GcRef::map(o.borrow(), |o| &o.data)`

Same with `GcRefMut`

Reply:
Ah right, we don't have `data_mut()`. Yeah, just add it to `Object<T>`

Reply:
Uh for some reason I missed `GcRef[Mut]::map`. Probably because it's still unstable in std. Thanks!